include/forcing/OptionalWrappedProvider.hpp

```diff
@@ -0,0 +1,361 @@
+                return false;
+            }
+
+            // Confirm this will provide anything
+            if (provider == nullptr) {
+                setMessage = "Cannot set provider as null, as this cannot provide the values this type must proxy";
+                return false;
+            }
+
+            // Confirm this will provide everything needed, though allow for defaults when available
+            for (const string &requiredName : providedOutputs) {
+                // When not provided by this provider and there is not a default value set up ...
+                if (!isSuppliedByProvider(requiredName, provider) && !isSuppliedWithDefault(requiredName)) {
+                    setMessage = "Given provider does not provide the required " + requiredName;
+                    return false;
+                }
+            }
+
+            // If this is good, set things and return true
+            wrapped_provider = provider;
+            setMessage.clear();
+            return true;
+        }
+
+    protected:
+
+        /**
+         * Record that there was an instance of a default value being used and manage the default usage "waits."
+         *
+         * When constructed, this instance may have been provided with a number of default usage "waits" for one or more
+         * outputs; i.e., a number of times it should wait to proxy a particular output value from the backing provider
+         * and instead allow a default value to override the aforementioned backing value.  This function, which should
+         * be called by @ref get_value any time a default is used, performs any modifications to the object's state that
+         * are necessary when a default is used, in particular with respect to managing "waits."
+         *
+         * In this base implementation, a recorded default usage does not count against the required number of "waits"
+         * for an output unless/until there is a wrapped backing provider set that can provide the given output.
+         * Otherwise, positive wait counts for this output be reduced by 1.  This allows negative wait counts to be used
+         * to represent that some output should always have the default used, even if the backing provider could supply
+         * a value for it.
+         *
+         * @param output_name The name of the output for which a default was used.
+         */
+        virtual void recordUsingDefault(const string &output_name) {
+            // Don't bother doing anything if there aren't waits assigned for this output
+            // Also, in this implementation, don't count usages until there is a backing provider that can provide this
+            auto waits_it = defaultUsageWaits.find(output_name);
+            if (waits_it != defaultUsageWaits.end() && isSuppliedByWrappedProvider(output_name)) {
+                // A value of 0 should be inferred and cleared from the collection, so ...
```

Comment:
So what exactly happens with a value of 0???

Reply:
@hellkite500, the intent was for there never be a value less than `1` in the `defaultUsageWaits` map.  If we assume this condition when an object is created, then it should hold throughout, as any values equal to 1 before the decrementing step are instead removed from the map entirely.

However, your comment prompted me to double-check and discover that it is possible for the `> 0` condition to be violated.  There is this constructor: 

```c++
OptionalWrappedProvider(vector<string> providedOuts, map<string, double> defaultVals, map<string, int> defaultWaits)
```

This validates the keys in `defaultWaits` (needed for other reasons) but not that the values are all greater than 0.

I could fix this, but first I'd like some opinions on which is the right approach.  Instead of what I was originally trying, does it make more sense to do as you had expected and check in the `recordUsingDefault` function for the `<= 0` case?  It's slightly less efficient but more intuitive.